File: KadirliApp.Domain/Entities/PowerOutage.cs
using KadirliApp.Domain.Common;

namespace KadirliApp.Domain.Entities;

public class PowerOutage : BaseEntity
{
    /// <summary>
    /// Faz 12.3'te <b>doldurulmaya başlandı</b>: kesinti bildirimi bir duyurudur ve bu alan
    /// o duyuruyu işaret eder. 10.x'ten beri var olan ama hiç yazılmayan çengeldi.
    /// </summary>
    public Guid? AnnouncementId { get; set; }

    /// <summary>
    /// 🔴 <b>TÜRETİLMİŞ ALAN (Faz 12.3'ten beri).</b> <see cref="NeighborhoodId"/> doluyken
    /// değer sözlükten (<see cref="Entities.Neighborhood.Name"/>) yazılır, elle düzenlenmez.
    /// </summary>
    /// <remarks>
    /// Kolon <b>bilerek duruyor ve DTO'daki adı değişmedi</b>: <c>GET /v1/power-outages</c>
    /// bu alanı düz metin olarak veriyor ve mağazadaki eski sürümler mahalle eşleşmesini
    /// <b>ad üzerinden</b> yapıyor (<c>power_outage.dart → matchesNeighborhood</c>). Kolonu
    /// kaldırmak ya da adını değiştirmek kırıcı bir değişiklik olurdu (<c>ARCHITECTURE.md</c> §5);
    /// türetmek ise additive: eski istemci aynı metni görmeye devam eder, üstelik artık
    /// yazım farkı olmadığı için eşleşme <b>daha</b> güvenilirdir.
    ///
    /// ⚠️ FK'sı olmayan (12.3 öncesinden kalan, geri doldurmada eşleşmemiş) kayıtlarda bu alan
    /// hâlâ serbest metindir — ve o kayıtlar <b>bildirim gönderemez</b>, çünkü hedeflenecek
    /// bir mahalle kimliği yoktur.
    /// </remarks>
    public string? Neighborhood { get; set; }

    /// <summary>
    /// Faz 12.3: kesintinin sözlükteki mahallesi. <c>null</c> = şehir geneli ya da henüz
    /// eşleşmemiş eski kayıt.
    /// </summary>
    public Guid? NeighborhoodId { get; set; }

    /// <summary>
    /// Faz 12.3: mahallenin hangi kısmı ("Atatürk Caddesi ve çevresi"). Kesinti çoğu zaman
    /// mahallenin tamamını kapsamıyor; bu ayrıntı önce serbest metin mahalle alanına
    /// sıkıştırılıyordu ve sözlük eşleşmesini imkânsız kılan asıl sebep oydu.
    /// </summary>
    public string? AreaDetail { get; set; }

    public DateTime StartTime { get; set; }
    public DateTime EndTime { get; set; }
    public string? Reason { get; set; }
    public string? Source { get; set; }

    public Announcement? Announcement { get; set; }
    public Neighborhood? NeighborhoodRef { get; set; }
}

File: mobile/lib/features/power_outages/data/models/power_outage.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'power_outage.freezed.dart';
part 'power_outage.g.dart';

/// `GET /v1/power-outages` öğesi (`PowerOutageDto`).
///
/// ⚠️ Uç **tüm** kayıtları döner (tarih filtresi yok, sayfalama yok) — geçmiş
/// kesintileri ayıklamak istemcinin işi ([isActive]/[isUpcoming]).
@freezed
abstract class PowerOutage with _$PowerOutage {
  const factory PowerOutage({
    required String id,

    /// Mahalle **adı**. Faz 12.3'ten beri sunucuda sözlükten türetiliyor (yazım
    /// farkı yok); eski kayıtlarda hâlâ serbest metin olabilir.
    String? neighborhood,

    /// Faz 12.3 (yeni): sözlükteki mahalle kimliği. Eski sürümlerde ve şehir
    /// geneli kesintilerde `null`.
    String? neighborhoodId,

    /// Faz 12.3 (yeni): mahallenin hangi kısmı ("Atatürk Caddesi ve çevresi").
    String? areaDetail,

    required DateTime startTime,
    required DateTime endTime,
    String? reason,

    /// Faz 12.3 (yeni): bu kesinti için üretilmiş duyuru. Dolu olması
    /// "bildirim gönderildi" demektir.
    String? announcementId,
  }) = _PowerOutage;

  const PowerOutage._();

  factory PowerOutage.fromJson(Map<String, dynamic> json) =>
      _$PowerOutageFromJson(json);

  /// Şu an sürüyor mu?
  bool isActive({DateTime? now}) {
    final reference = (now ?? DateTime.now()).toUtc();
    return !startTime.toUtc().isAfter(reference) && endTime.toUtc().isAfter(reference);
  }

  /// Henüz başlamadı mı (planlı)?
  bool isUpcoming({DateTime? now}) =>
      startTime.toUtc().isAfter((now ?? DateTime.now()).toUtc());

  /// Ana Sayfa şeridinde gösterilmeye değer mi (süren ya da gelecek)?
  bool isRelevant({DateTime? now}) => isActive(now: now) || isUpcoming(now: now);

  /// Bitmiş mi?
  bool isPast({DateTime? now}) =>
      !endTime.toUtc().isAfter((now ?? DateTime.now()).toUtc());

  /// Üç durumdan biri — liste gruplaması ve rozet buradan okunur.
  PowerOutageStatus status({DateTime? now}) {
    if (isActive(now: now)) return PowerOutageStatus.active;
    if (isUpcoming(now: now)) return PowerOutageStatus.upcoming;
    return PowerOutageStatus.past;
  }

  /// Planlanan toplam kesinti süresi.
  Duration get duration => endTime.difference(startTime);

  /// Süren kesintide bitişe, planlıda başlangıca kalan süre; geçmişte null.
  Duration? remaining({DateTime? now}) {
    final reference = (now ?? DateTime.now()).toUtc();
    return switch (status(now: reference)) {
      PowerOutageStatus.active => endTime.toUtc().difference(reference),
      PowerOutageStatus.upcoming => startTime.toUtc().difference(reference),
      PowerOutageStatus.past => null,
    };
  }

  /// Mahalle adı — boş/whitespace gelirse şehir geneli sayılır.
  String get placeLabel {
    final name = neighborhood?.trim();
    return (name == null || name.isEmpty) ? 'Kadirli geneli' : name;
  }

  /// Bu kesinti mahalle sözlüğüne bağlı mı (Faz 12.3).
  bool get hasNeighborhoodRef => (neighborhoodId?.trim().isNotEmpty ?? false);

  /// Mahalle bilgisi hiç yok → şehir geneli, herkesi ilgilendirir.
  bool get isCityWide => (neighborhood?.trim().isEmpty ?? true);

  /// Kullanıcının mahallesiyle eşleşiyor mu.
  ///
  /// 🔑 Faz 12.3: **önce kimlik, sonra ad.** Uç 12.3'ten beri `neighborhoodId`
  /// döndürüyor; kimlik varsa ad karşılaştırması hiç yapılmaz — "Cengiz Topel"
  /// ile "Cengiz Topel Mahallesi" yazım farkı yüzünden sessizce eşleşmeyen
  /// kayıtlar tam olarak buradan doğuyordu.
  ///
  /// ⚠️ Ad karşılaştırması **kaldırılmadı**: mağazadaki eski sürümlerin değil,
  /// sunucudaki eski *kayıtların* hatırına — geri doldurmada eşleşmemiş bir
  /// kesintinin `neighborhoodId`'si hâlâ boş gelir ve o kayıt için elimizdeki
  /// tek şey ad.
  bool matchesNeighborhood(String? userNeighborhood, {String? userNeighborhoodId}) {
    final myId = userNeighborhoodId?.trim();
    final hereId = neighborhoodId?.trim();
    if (myId != null && myId.isNotEmpty && hereId != null && hereId.isNotEmpty) {
      return myId == hereId;
    }

    final mine = userNeighborhood?.trim().toLowerCase();
    final here = neighborhood?.trim().toLowerCase();
    if (mine == null || mine.isEmpty || here == null || here.isEmpty) {
      return false;
    }
    return mine == here;
  }
}

/// Kesintinin zamana göre durumu.
enum PowerOutageStatus { active, upcoming, past }
